negotiator.go

```
@@ -149,3 +181,46 @@ func (l Negotiator) RoundTrip(req *http.Request) (res *http.Response, err error)
 
 	return res, err
 }
+
+func makeChannelBinding(state tls.ConnectionState) ([]byte, error) {
+
+	certificate := state.PeerCertificates[0]
+	prefix := []byte("tls-server-end-point:")
+
+	if certificate == nil {
+		return nil, errors.New("TLS connection is missing server certificate")
+	}
+
+	// choose the channel binding hash type
+	// Use the same hash type used for the certificate signature, except for MD5 and SHA-1 which
+	// use SHA256
+	hashType := crypto.SHA256
+	switch certificate.SignatureAlgorithm {
+	case x509.SHA384WithRSA, x509.ECDSAWithSHA384, x509.SHA384WithRSAPSS:
+		hashType = crypto.SHA384
+	case x509.SHA512WithRSA, x509.ECDSAWithSHA512, x509.SHA512WithRSAPSS:
+		hashType = crypto.SHA512
+	}
+
+	hasher := hashType.New()
+	_, _ = hasher.Write(certificate.Raw)
+	data := hasher.Sum(nil)
+
+	buf := bytes.NewBuffer(make([]byte, 0, len(prefix)+len(data)))
+	buf.Write(prefix)
```

if we ignore errors we need a comment explaining why in the code